### ISSUE TITLE
Graceful shutdown

### ISSUE THREAD

**Issue by burdiyan** [hsanjuan/ipfs-lite#12]
I was looking at the code and something caught my attention. I see that instantiating a peer would spin up a goroutine that would wait on the context.Done channel and then close some of the underlying dependencies.

I might be wrong, but it seems like there's no way to wait until this shutdown process is actually finished. The `autoclose` goroutine is not signaling anywhere when it's done.

Wouldn't it be more correct to provide explicit `Close` method so that users of ipfs-lite could chose to wait until the peer is actually closed? Another approach could be providing an explicit `Start` method that would block and could accept a context for being canceled.

**Comment by hsanjuan** [comment]
> Wouldn't it be more correct to provide explicit Close method so that users of ipfs-lite could chose to wait until the peer is actually closed?

Yes, probably better. Do you have a situation where this matters or generates a problem?

**Comment by burdiyan** [comment]
Not really. I'm just a bit of a graceful shutdown freak to be honest :)

**Comment by hsanjuan** [comment]
I'll keep this open. Things do break sometimes because of this, and I think go-ipfs has issues with the reprovider being closed after the datastore and things like that.

**Comment by burdiyan** [comment]
I'd submit a PR, but probably you'd have to cut a v2 release for this change since it wouldn't be backward compatible. Unless we add an explicit Close method that would wait until autoclose goroutine returns. WDYT?

**Comment by hsanjuan** [comment]
AutoClose can be an option enabled by default, and the Close method added additionally.
